src/tdigest/sketch.rs

```diff
@@ -0,0 +1,1138 @@
+            if preamble_longs == 0 && serial_version == 0 && family_id == 0 {
+                return Self::deserialize_compat(bytes);
+            }
+            return Err(SerdeError::InvalidFamily(format!(
+                "expected {} (TDigest), got {}",
+                TDIGEST_FAMILY_ID, family_id
+            )));
+        }
+        if serial_version != SERIAL_VERSION {
+            return Err(SerdeError::UnsupportedVersion(format!(
+                "expected {}, got {}",
+                SERIAL_VERSION, serial_version
+            )));
+        }
+        let k = cursor.read_u16::<LE>().map_err(make_error("k"))?;
+        let flags = cursor.read_u8().map_err(make_error("flags"))?;
+        let is_empty = (flags & FLAGS_IS_EMPTY) != 0;
+        let is_single_value = (flags & FLAGS_IS_SINGLE_VALUE) != 0;
+        let expected_preamble_longs = if is_empty || is_single_value {
+            PREAMBLE_LONGS_EMPTY_OR_SINGLE
+        } else {
+            PREAMBLE_LONGS_MULTIPLE
+        };
+        if preamble_longs != expected_preamble_longs {
+            return Err(SerdeError::MalformedData(format!(
+                "expected preamble_longs to be {}, got {}",
+                expected_preamble_longs, preamble_longs
+            )));
+        }
+        cursor.read_u16::<LE>().map_err(make_error("<unused>"))?; // unused
+        if is_empty {
+            return Ok(TDigestMut::new(k));
+        }
+
+        let reverse_merge = (flags & FLAGS_REVERSE_MERGE) != 0;
+        if is_single_value {
+            let value = if is_f32 {
+                cursor
+                    .read_f32::<LE>()
+                    .map_err(make_error("single_value"))? as f64
+            } else {
+                cursor
+                    .read_f64::<LE>()
+                    .map_err(make_error("single_value"))?
+            };
+            check_non_nan(value, "single_value")?;
+            return Ok(TDigestMut::make(
+                k,
+                reverse_merge,
+                value,
+                value,
+                vec![Centroid {
+                    mean: value,
+                    weight: 1,
+                }],
+                1,
+                vec![],
+            ));
+        }
+        let num_centroids = cursor
+            .read_u32::<LE>()
+            .map_err(make_error("num_centroids"))? as usize;
+        let num_buffered = cursor
+            .read_u32::<LE>()
+            .map_err(make_error("num_buffered"))? as usize;
+        let (min, max) = if is_f32 {
+            (
+                cursor.read_f32::<LE>().map_err(make_error("min"))? as f64,
+                cursor.read_f32::<LE>().map_err(make_error("max"))? as f64,
+            )
+        } else {
+            (
+                cursor.read_f64::<LE>().map_err(make_error("min"))?,
+                cursor.read_f64::<LE>().map_err(make_error("max"))?,
+            )
+        };
+        check_non_nan(min, "min")?;
+        check_non_nan(max, "max")?;
```

Comment:
min & max should have other certia like min <= max and should be consistent with centroid + buffer. But I'd leave it later since the non_nan check is complex enough here. I don't make this PR too unreviewable (